ethreceipts/ethreceipts.go

```diff
@@ -0,0 +1,197 @@
+	// again, or it might re-price a txn, or other.
+	WaitNumBlocksBeforeExhaustion int
+}
+
+var DefaultOptions = Options{
+	NumBlocksUntilTxnFinality:     12,
+	WaitNumBlocksBeforeExhaustion: 3,
+}
+
+type Event uint32
+
+// We might not need Unknown here, we can use Discarded instead.
+const (
+	Finalized Event = iota
+	Dropped
+	Unknown
+)
+
+var (
+	ErrExhausted = errors.New("ethreceipts: exhausted looking for receipt")
+)
+
+func NewReceipts(log zerolog.Logger, provider *ethrpc.Provider, monitor *ethmonitor.Monitor, opts ...Options) (*Receipts, error) {
+	options := DefaultOptions
+	if len(opts) > 0 {
+		options = opts[0]
+	}
+
+	if options.NumBlocksUntilTxnFinality <= 0 {
+		return nil, fmt.Errorf("ethreceipts: invalid option, NumBlocksUntilTxnFinality")
+	}
+	if options.WaitNumBlocksBeforeExhaustion <= 2 {
+		return nil, fmt.Errorf("ethreceipts: invalid option, WaitNumBlocksBeforeExhaustion")
+	}
+
+	return &Receipts{
+		options:  options,
+		log:      log.With().Str("ps", "receipts").Logger(),
+		provider: provider,
+		mointor:  monitor,
+	}, nil
+}
+
+func (r *Receipts) Run(ctx context.Context) error {
+	if r.IsRunning() {
+		return fmt.Errorf("ethreceipts: already running")
+	}
+
+	r.ctx, r.ctxStop = context.WithCancel(ctx)
+
+	atomic.StoreInt32(&r.running, 1)
+	defer atomic.StoreInt32(&r.running, 0)
+
+	// NOTE: this first version new of ethreceipts doesn't have a sub-process running. The Run`ner
+	// is just mocked here for when we do something more advanced described below with ethmonitor.
+	<-r.ctx.Done()
+
+	// TODO: so, the current version of ethreceipts will poll for the receipts. However,
+	// this will be a ton of http requests to the node coming from this service. It's fine for now,
+	// but in the next iteration we can use a single ethmonitor instance and listen for txn hashes
+	// as they are mined, and then fetch the receipt afterwards once we know it'll be available for sure. (Done)
+	// We can also use the monitor's cache and block retention to look for historical receipts
+	// up to a certain limit too.
+
+	return nil
+}
+
+func (r *Receipts) Stop() {
+	if !r.IsRunning() {
+		return
+	}
+	r.log.Info().Str("op", "stop").Msg("-> receipts: stopping..")
+	r.ctxStop()
+	r.log.Info().Str("op", "stop").Msg("-> receipts: stopped")
+}
+
+func (r *Receipts) IsRunning() bool {
+	return atomic.LoadInt32(&r.running) == 1
+}
+
+func (r *Receipts) Options() Options {
+	return r.options
+}
+
+// GetTransactionReceipt is a blocking operation that will listen for chain blocks looking for the txn hash
+// provided and will then respond with the receipt.
+func (r *Receipts) GetTransactionReceipt(ctx context.Context, txnHash common.Hash) (*types.Receipt, error) {
+	startTime := time.Now()
+
+	for {
+		select {
+		case <-ctx.Done():
+			if err := ctx.Err(); err != nil {
+				return nil, fmt.Errorf("%w: unable to find %v: %v", ErrExhausted, txnHash.Hex(), err)
+			}
+
+		default:
+			tx := r.mointor.GetTransaction(txnHash)
```

Comment:
few issues..

1. `mointor` is a typo
2. this method is non-blocking, so this for {} loop is going to eat CPU is an almost infinite loop -- bad.